app/Services/InboundMail/InboundMailEngine.php

```diff
@@ -0,0 +1,272 @@
+        }
+
+        return false;
+    }
+    public function blockSender(string $from)
+    {
+        Cache::add('inboundMailBlockedSender:' . $from, true, now()->addHours(12));
+
+        // TODO: ignore, when known sender (for heavy email-usage mostly on isHosted())
+        // TODO: handle external blocking
+    }
+    public function saveMeta(string $from, string $to, bool $isUnknownRecipent = false)
+    {
+        // save cache
+        Cache::add('inboundMailCountSender:' . $from, 0, now()->addHours(12));
+        Cache::increment('inboundMailCountSender:' . $from);
+
+        if ($isUnknownRecipent) {
+            Cache::add('inboundMailCountSenderUnknownRecipent:' . $from, 0, now()->addHours(6));
+            Cache::increment('inboundMailCountSenderUnknownRecipent:' . $from); // we save the sender, to may block him
+
+            Cache::add('inboundMailCountUnknownRecipent:' . $to, 0, now()->addHours(12));
+            Cache::increment('inboundMailCountUnknownRecipent:' . $to); // we save the sender, to may block him
+        }
+    }
+
+    // MAIN-PROCESSORS
+    protected function createExpense(Company $company, InboundMail $email)
+    {
+        // Skipping executions: will not result in not saving Metadata to prevent usage of these conditions, to spam
+        if (!($company?->expense_mailbox_active ?: false)) {
+            $this->logBlocked($company, 'mailbox not active for this company. from: ' . $email->from);
+            return;
+        }
+        if (!$this->validateExpenseSender($company, $email)) {
+            $this->logBlocked($company, 'invalid sender of an ingest email for this company. from: ' . $email->from);
+            return;
+        }
+        if (sizeOf($email->documents) == 0) {
+            $this->logBlocked($company, 'email does not contain any attachments and is likly not an expense. from: ' . $email->from);
+            return;
+        }
+
+        // create expense
+        $expense = ExpenseFactory::create($company->id, $company->owner()->id);
+
+        $expense->public_notes = $email->subject;
+        $expense->private_notes = $email->text_body;
+        $expense->date = $email->date;
+
+        // handle vendor assignment
+        $expense_vendor = $this->getVendor($company, $email);
+        if ($expense_vendor)
+            $expense->vendor_id = $expense_vendor->id;
+
+        // handle documents
+        $this->processHtmlBodyToDocument($email);
+        $documents = [];
+        array_push($documents, ...$email->documents);
+        if ($email->body_document !== null)
+            array_push($documents, $email->body_document);
+
+        $expense->saveQuietly();
+
+        $this->saveDocuments($documents, $expense);
+
+        event(new ExpenseWasCreated($expense, $expense->company, Ninja::eventVars(null))); // @turbo124 please check, I copied from API-Controller
+        event('eloquent.created: App\Models\Expense', $expense); // @turbo124 please check, I copied from API-Controller
```

Comment:
@turbo124 when creating an expense through the api the following events are thrown:

event(new ExpenseWasCreated($expense, $expense->company, Ninja::eventVars(null))); // @turbo124 please check, I copied from API-Controller
event('eloquent.created: App\Models\Expense', $expense); // @turbo124 please check, I copied from API-Controller
should they also be present here?